packages/contracts-bedrock/scripts/L2Genesis.s.sol

```diff
@@ -0,0 +1,381 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.15;
+
+import { Script } from "forge-std/Script.sol";
+import { console2 as console } from "forge-std/console2.sol";
+
+import { Artifacts } from "scripts/Artifacts.s.sol";
+import { DeployConfig } from "scripts/DeployConfig.s.sol";
+import { Predeploys } from "src/libraries/Predeploys.sol";
+import { L1StandardBridge } from "src/L1/L1StandardBridge.sol";
+import { L1CrossDomainMessenger } from "src/L1/L1CrossDomainMessenger.sol";
+import { L2StandardBridge } from "src/L2/L2StandardBridge.sol";
+import { L2CrossDomainMessenger } from "src/L2/L2CrossDomainMessenger.sol";
+import { SequencerFeeVault } from "src/L2/SequencerFeeVault.sol";
+import { FeeVault } from "src/universal/FeeVault.sol";
+import { OptimismMintableERC20Factory } from "src/universal/OptimismMintableERC20Factory.sol";
+import { L1Block } from "src/L2/L1Block.sol";
+import { GovernanceToken } from "src/governance/GovernanceToken.sol";
+import { Strings } from "@openzeppelin/contracts/utils/Strings.sol";
+
+interface IInitializable {
+    function initialize(address _addr) external;
+}
+
+/// @dev The general flow of adding a predeploy is:
+///      1. _setProxies uses vm.etch to set the Proxy.sol deployed bytecode for proxy address 0x420...0 to 0x420...799.
+///      Additionally, the PROXY_ADMIN_ADDRESS and PROXY_IMPLEMENTATION_ADDRESS storage slots are set for the proxy address.
+///      2. vm.etch sets the deployed bytecode for each predeploy at the implementation address (i.e. code namespace)
+///      3. The `initialize` method is called at the implementation address with zero/dummy vaules if it exists
+///      4. The `initialize` method is called at the proxy address with actual vaules if it exists
+///      5. A `require` check to verify the expected implementation address is set for the proxy
+contract L2Genesis is Script, Artifacts {
+    uint256 constant PROXY_COUNT = 2048;
+    uint256 constant PRECOMPILE_COUNT = 256;
+    DeployConfig public constant cfg =
+        DeployConfig(address(uint160(uint256(keccak256(abi.encode("optimism.deployconfig"))))));
+
+    /// @notice The storage slot that holds the address of a proxy implementation.
+    /// @dev `bytes32(uint256(keccak256('eip1967.proxy.implementation')) - 1)`
+    bytes32 internal constant PROXY_IMPLEMENTATION_ADDRESS =
+        0x360894a13ba1a3210667c828492db98dca3e2076cc3735a920a3ca505d382bbc;
+
+    /// @notice The storage slot that holds the address of the owner.
+    /// @dev `bytes32(uint256(keccak256('eip1967.proxy.admin')) - 1)`
+    bytes32 internal constant PROXY_ADMIN_ADDRESS = 0xb53127684a568b3173ae13b9f8a6016e243e63b6e8ee1178d6a717850b5d6103;
+
+    string internal outfile;
+
+    function setUp() public override {
+        Artifacts.setUp();
+
+        string memory path = string.concat(vm.projectRoot(), "/deploy-config/", deploymentContext, ".json");
```

Comment:
This pattern is very common, and i would like to deduplicate the code, will follow up with a plan eventually, likely a `setUp` function inside of `DeployConfig` could reduce this boilerplate, dont need to worry about it now

Reply:
Tracking via this issue